File: range/rangeminqueries1.cc
#include <bits/stdc++.h>
using namespace std;

int li[200001], st[800001];

void build(int si, int ss, int se)
{
    if (ss == se)
    {
        st[si] = li[ss];
        return;
    }
    int mid = (ss + se) / 2;
    build(2 * si, ss, mid);
    build(2 * si + 1, mid + 1, se);
    st[si] = min(st[2 * si], st[2 * si + 1]);
}

int query(int si, int ss, int se, int qs, int qe)
{
    if (qs > se || qe < ss)
    {
        return INT_MAX;
    }
    if (ss >= qs && se <= qe)
    {
        return st[si];
    }
    int mid = (ss + se) / 2;
    int l = query(2 * si, ss, mid, qs, qe);
    int r = query(2 * si + 1, mid + 1, se, qs, qe);
    return min(l, r);
}

int main()
{
    int n, q;
    cin >> n >> q;
    for (int o = 1; o <= n; o++)
        cin >> li[o];
    build(1, 1, n);
    while (q--)
    {
        int a, b;
        cin >> a >> b;
        cout << query(1, 1, n, a, b) << '\n';
    }
    return 0;
}

File: range/rangeminqueries1.py
from sys import stdin, maxsize
# li = [0] * 1000000001  # Our array
# st = [0] * 4000000001  # Our Segment Tree

# Segment tree array is indexed 1 based


def build(li, st, si, ss, se):
    # si = segment index
    # ss = segment start
    # se = segment end
    if ss == se:
        st[si] = li[ss]
        return

    mid = (ss + se) // 2
    build(li, st, 2 * si, ss, mid)
    build(li, st, 2 * si + 1, mid + 1, se)
    st[si] = min(st[2 * si], st[2 * si + 1])


def query(st, si, ss, se, qs, qe):
    # qs = query start
    # qe = query end
    if ss > qe or se < qs:
        return maxsize  # Completely out of the segment and doesn't contribute to the solution

    if ss >= qs and se <= qe:
        return st[si]   # Lies completely in the segment
    mid = (ss + se) // 2
    # Return minimum from left and right subtree
    return min(query(st, 2 * si, ss, mid, qs, qe), query(st, 2 * si + 1, mid + 1, se, qs, qe))


def main():
    n, q = map(int, stdin.readline().split())
    li = [0] * (n + 1)
    st = [0] * (4 * (n + 1))
    c = 1
    for i in stdin.readline().split():
        li[c] = int(i)
        c += 1
    build(li, st, 1, 1, n)
    while q:
        a, b = map(int, stdin.readline().split())
        print(query(st, 1, 1, n, a, b))
        q -= 1


if __name__ == "__main__":
    main()
